src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Cohost/DidChangeHandler.cs

```diff
@@ -14,15 +15,41 @@ namespace Microsoft.VisualStudio.LanguageServerClient.Razor.Cohost;
 
 [Export(typeof(IRazorCohostDidChangeHandler)), Shared]
 [method: ImportingConstructor]
-internal class DidChangeHandler(ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher, RazorProjectService razorProjectService) : IRazorCohostDidChangeHandler
+internal class DidChangeHandler(
+    ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher,
+    RazorProjectService razorProjectService,
+    ISnapshotResolver snapshotResolver,
+    OpenDocumentGenerator openDocumentGenerator) : IRazorCohostDidChangeHandler
 {
     private readonly ProjectSnapshotManagerDispatcher _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
     private readonly RazorProjectService _razorProjectService = razorProjectService;
+    private readonly ISnapshotResolver _snapshotResolver = snapshotResolver;
+    private readonly OpenDocumentGenerator _openDocumentGenerator = openDocumentGenerator;
 
-    public Task HandleAsync(Uri uri, int version, SourceText sourceText, CancellationToken cancellationToken)
+    public async Task HandleAsync(Uri uri, int version, SourceText sourceText, CancellationToken cancellationToken)
     {
-        return _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(
-            () => _razorProjectService.UpdateDocument(uri.GetAbsoluteOrUNCPath(), sourceText, version),
-            cancellationToken);
+        await await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(async () =>
+        {
+            var textDocumentPath = FilePathNormalizer.Normalize(uri.GetAbsoluteOrUNCPath());
+            _razorProjectService.UpdateDocument(textDocumentPath, sourceText, version);
+
+            // We are purposefully trigger code generation here directly, rather than using the project manager events that the above call
+            // would have triggered, because Cohosting is intended to eventually remove the project manager and its events. We also want
+            // to eventually remove this code too, and just rely on the source generator, but by keeping the concepts separate we are not
+            // tying the code to any particular order of feature removal.
+            if (!_snapshotResolver.TryResolveAllProjects(textDocumentPath, out var projectSnapshots))
+            {
+                projectSnapshots = [_snapshotResolver.GetMiscellaneousProject()];
+            }
+
+            foreach (var project in projectSnapshots)
+            {
+                var document = project.GetDocument(textDocumentPath);
+                if (document is not null)
+                {
+                    await _openDocumentGenerator.DocumentOpenedOrChangedAsync(document, version, cancellationToken);
+                }
+            }
```

Comment:
Nit: this part seems to be the same in DidOpenHandler, not sure if it makes sense to create some sort of a helper to re-use the code. If we will be removing features down the road and modifying it, might be nice to do it in one place. Feel free to punt, not a big deal.

Reply:
Yep, good call, will move to OpenDocumentGenerator